doc/horizontal_autoscaling.md

```diff
@@ -0,0 +1,258 @@
+
+
+### Training Job Controller
+
+Currently, we will run a single training job controller instance and
+assume that there is no training job controller running concurrently
+(the assumption could be false
+when
+[split-brain](https://en.wikipedia.org/wiki/Split-brain_(computing))
+happens). In the future, we will run multiple instances and use leader
+election to choose a leader.
+
+The pseudo-logic is as follows:
+
+```go
+registerThirdPartyResource()
+for {
+  quota := getTotalComputationResourceQuota()
+  current := getCurrentJobStates()
+  desired := getDesiredJobStates()
+  dynamicScaling(quota, current, desired)
+}
+```
+
+#### Scaling Algorithm
+
+##### Elastic Job
+
+A job is elastic only when it's trainer and pserver's `min-instance`
+equals to the `max-instance` respectively. We will only scale elastic
+jobs.
+
+Currently, we will not scale the parameter server instances.
+
+##### Fulfillment Score
+
+When there are available computation resources, the algorithm needs to
+decide which jobs to assign the resources to. When there are no more
+available computation resources but the newly submitted job needs it,
+the algorithm needs to decide which job to take the resource away
+from. We will introduce the *fulfillment score* to answer these
+questions:
+
+```go
+func (j Job) Score() float64 {
+  minInstance := j.spec.trainer.minInstance
+  maxInstance := j.spec.trainer.minInstance
+  curInstance := j.trainer.currentInstance()
+  return float64(curInstance - minInstance) / float64(maxInstance - minInstance)
+}
+```
+
+##### Scaling GPU Jobs
+
+The controller knows the total number of available GPUs in a cluster
+and will try to assign all of them to the training jobs.
+
+All elastic GPU jobs will be sorted according to their fulfillment
+score. The number of GPU per instance, CPU requests value, Mem
+requests value will be used as tiebreakers in decreasing importance.
+
+An available GPU resource will be assigned to the least fulfilled job
+unless that job is already fulfilled (with a fulfillment score of
+`1.0`). A GPU resource will be take away from the most fulfilled job
+when there is another GPU job's `min-instance` is not satisfied
+(unless the most fulfilled job's `cur-instance` equals to
+`min-instance`). When the most fulfilled job's `cur-instance` equals
+to `min-instance`, no training job will be scaled down, the new job
+cannot be scheduled and will wait for more resources.
+
+
+##### Scaling CPU Jobs
+
+The controller knows the total CPU capacity, Mem capacity of the
+cluster, and the total CPU limits, Mem limits of all training jobs. We
+define the available CPU and Mem as the difference of the capacity and
+the
+[limits](https://kubernetes.io/docs/concepts/policy/resource-quotas/#requests-vs-limits) (not
+the
+[requests](https://kubernetes.io/docs/concepts/policy/resource-quotas/#requests-vs-limits))
+respectively.
+
+All elastic CPU jobs will be sorted according to their fulfillment
```

Comment:
Fulfillment score is awesome. Maybe it's not enough just use it. 
Do we need to classify the jobs by level and high-level tasks are prioritized?